lib/irq.h

```diff
@@ -95,6 +143,158 @@ void metal_irq_enable(unsigned int vector);
  */
 void metal_irq_disable(unsigned int vector);
 
+#define METAL_IRQ_CONTROLLER_DECLARE(_irq_controller, \
+				     _irq_base, _irq_num, \
+				     _arg, \
+				     _irq_set_state, \
+				     _irqs) \
+	struct metal_irq_controller _irq_controller = { \
+		.irq_base = _irq_base, \
+		.irq_num = _irq_num, \
+		.arg = _arg, \
+		.irq_set_state = _irq_set_state, \
+		.irqs = _irqs,\
+	};
+
+/**
+ * @brief	metal_irq_register_controller
+ *
+ * Register IRQ controller
+ *
+ * @param[in] cntr  Interrupt controller to register
+ * @return 0 on success, or negative value for failure.
+ */
+void metal_irq_register_controller(struct metal_irq_controller *cntr);
+
+/**
+ * @brief	metal_irq_allocate
+ *
+ * Allocate IRQ range
+ *
+ * @param[in] irq_base Base of required IRQ range, it can be METAL_IRQ_ANY.
+ *                     If it is METAL_IRQ_ANY, libmetal will allocate the
+ *                     base or IRQ range.
+ * @param[in] irq_num  Number of IRQs required
+ * @return irq number base for success, or negative value for failure.
+ */
+int metal_irq_allocate(int irq_base, int irq_num);
+
+/**
+ * @brief	metal_irq_map_softirq
+ *
+ * Map soft IRQ range with input IRQ range which will trigger the soft
+ * IRQs.
+ *
+ * @param[in] source_irq_base Base of source IRQ range.
+ * @param[in] irq_num  Number of IRQs required
+ * @return base of target soft IRQs for success, or negative value for
+ *         failure.
+ */
+int metal_irq_map_softirq(int source_irq_base, int irq_num);
```

Comment:
I don't think we should have a mapping from HW irq to soft irq since there is no guarantee that there is a 1:1 mapping.  often there is no mapping and sometimes there is a 1 to many mapping. I think this should be handled by the user.

Reply:
Agree, will remove this from irq.h